src/NUnitConsole/nunit-console/Utilities/TeamCityServiceMessagePublisher.cs

```diff
@@ -0,0 +1,389 @@
+﻿// ***********************************************************************
+// Copyright (c) 2015 Charlie Poole
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+// 
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+using System.IO;
+using System.Xml;
+using System.Collections.Generic;
+using System.Globalization;
+
+namespace NUnit.ConsoleRunner.Utilities
+{
+    using System;
+
+    internal class TeamCityServiceMessagePublisher
+    {
+        private readonly TextWriter _outWriter;
+        private readonly Dictionary<string, string> _refs = new Dictionary<string, string>();
+        private int _blockCounter;
+        private string _rootFlowId;
+
+        public TeamCityServiceMessagePublisher(TextWriter outWriter)
+        {
+            if (outWriter == null)
+            {
+                throw new ArgumentNullException("outWriter");
+            }
+
+            _outWriter = outWriter;
+        }
+
+        public void RegisterMessage(XmlNode message)
+        {
+            if (message == null)
+            {
+                throw new ArgumentNullException("message");
+            }
+
+            var messageName = message.Name;
+            if (string.IsNullOrEmpty(messageName))
+            {
+                return;
+            }
+
+            messageName = messageName.ToLowerInvariant();
+            if (messageName == "start-run")
+            {
+                _refs.Clear();
+                return;
+            }           
+
+            var fullName = message.GetAttribute("fullname");
+            if (string.IsNullOrEmpty(fullName))
+            {
+                return;
+            }
+
+            var id = message.GetAttribute("id");
+            var parentId = message.GetAttribute("parentId");
```

Comment:
Two things here...

1. What do you do with the hierarchy? In TeamCity reports I have seen, it's just a list of test cases. Will this be something new?

2. In the NUnit Gui, we need the hierarchy, obviously. We get it by first exploring the tests. The gui keeps the full hierarchy available. When any result comes in, it's easy to determine where it goes in the hierarchy. Is this a possible approach for you? Or do you find having it in each individual result is better? I'd like to hear the reasoning because maybe we want to do the same for the **new** Gui. :-)